### PR TITLE
Services

### PR DESCRIPTION
**Pull request by johnwhitton**
Updated DC and Tweet contracts

### CODE REVIEW

File: contracts/contracts/DC.sol [hunk 1]
```diff
@@ -6,135 +6,71 @@ import "@openzeppelin/contracts/access/Ownable.sol";
...
     bool public initialized;
 
     struct InitConfiguration {
-        uint256 baseRentalPrice;
-        uint256 duration;
-        uint256 gracePeriod;
 
-        // 32-bytes block
-        address revenueAccount;
+        // 12-bytes block
```
Comment (polymorpher) [inline contracts/contracts/DC.sol]: these fields should be aligned to 32-byte blocks or the alignment won't be meaningful at all

File: contracts/README.md [hunk 2]
```diff
@@ -6,4 +6,67 @@ DC contracts have compatible NameRecord structures from [D1DC (.1.country)](http
...
 3. The DC contract's companion frontend (in `/client` folder) is meant to be used together so the user can rent a web3 and web2 domain simutaneously. 
 
-In other words, the DC contract is meant to provide some "extra service" on top of ENS contracts. Therefore, it charges some additional fees on top of purchasing an ENS domain from the ENS app. On a product level, the main value is to host and configure data that is also viewable in web2. The initial product only allows the renter to embed a tweet on their new domain, but more functionalities are coming soon
\ No newline at end of file
+In other words, the DC contract is meant to provide some "extra service" on top of ENS contracts. Therefore, it charges some additional fees on top of purchasing an ENS domain from the ENS app. On a product level, the main value is to host and configure data that is also viewable in web2. The initial product only allows the renter to embed a tweet on their new domain, but more functionalities are coming soon
+
+
```
Comment (polymorpher) [inline contracts/README.md:11]: README is not TODO. Move dev notes to dev folder

File: contracts/contracts/DC.sol (hunk 1 above)
Comment (johnwhitton) [reply contracts/contracts/DC.sol]: ok, I have updated (addresses are 20 so I commented it that way, feel free to change as preferred)

File: contracts/README.md (hunk 2 above)
Comment (johnwhitton) [reply contracts/README.md:11]: Moved to dev/metadata.md

File: contracts/contracts/Tweet.sol [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+            chargeableDuration = expiration - block.timestamp;
+        }
+        if (expiration > nameRecord.expirationTime) {
+            chargeableDuration = expiration - nameRecord.expirationTime;
+        }
+        uint256 charge = (chargeableDuration * 1e18 / duration * baseRentalPrice) / 1e18;
+        return charge;
+    }
+
+    function reinstate(string calldata name) public payable whenNotPaused {
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:267]: not needed

File: contracts/contracts/Tweet.sol [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    }
+
+    /**
+     * @dev `renew` calls RegistrarController renew and is used to renew a name
+     * this also takes a fee for the web2 renewal which is held by DC.sol a check is made to ensure the value sent is sufficient for both fees
+     * duration is set at the contract level
+     * @param name The name to be registered e.g. for test.country it would be test
+     * @param url A URL that can be embedded in a web2 default domain page e.g. a twitter post
+     */
+    function renew(string calldata name, string calldata url) public payable whenNotPaused {
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:224]: not needed

File: contracts/contracts/Tweet.sol [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+
+    /**
+     * @dev `register` calls RegistrarController register and is used to register a name
+     * this also takes a fee for the web2 registration which is held by DC.sol a check is made to ensure the value sent is sufficient for both fees
+     * @param name The name to be registered e.g. for test.country it would be test
+     # @param owner The owner of the registerd name
+     * @param url A URL that can be embedded in a web2 default domain page e.g. a twitter post
+     * @param secret A random secret passed by the client
+     */
+    function register(string calldata name, address owner, string calldata url, bytes32 secret) public payable whenNotPaused {
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:188]: not needed

File: contracts/contracts/Tweet.sol [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+        _unpause();
+    }
+
+
+    function numRecords() public view returns (uint256){
+        return keys.length;
+    }
+
+    function getRecordKeys(uint256 start, uint256 end) public view returns (bytes32[] memory){
+        require(end > start, "D1DC: end must be greater than start");
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:131]: wrong message label

File: contracts/contracts/Tweet.sol [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+    uint256 public gracePeriod;
+    uint256 public baseRentalPrice;
+    address public revenueAccount;
+    uint256 public duration;
+    IDC     public dc;
+    bool public initialized;
+
+    struct InitConfiguration {
+        uint256 baseRentalPrice;
+        uint256 duration;
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:30]: property for DC

File: contracts/contracts/Tweet.sol [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+    }
+
+    /**
+     * @dev `available` calls RegistrarController to check if a name is available
+     * @param name The name to be checked being registered
+     */
+    function available(string memory name) public view returns (bool) {
+        NameRecord storage record = nameRecords[keccak256(bytes(name))];
+        bool ensAvailable = dc.available(name);
+        return ensAvailable && (record.renter == address(0) || uint256(record.expirationTime) + gracePeriod <= block.timestamp);
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:146]: no renter

File: contracts/contracts/Tweet.sol [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+    }
+
+    /**
+     * @dev `makeCommitment` calls RegistrarController makeCommitment with pre-populated values
+     * commitment is just a keccak256 hash
+     * @param name The name being registered
+     * @param owner The address of the owner of the name being registered
+     * @param secret A random secret passed by the client
+     */
+    function makeCommitment(string memory name, address owner, bytes32 secret) public view returns (bytes32) {
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:156]: not needed

File: contracts/contracts/Tweet.sol [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+        return dc.makeCommitment(name, owner, duration, secret);
+    }
+
+    /**
+     * @dev `commitment` calls RegistrarController commitment and is used as a locker to ensure that only one registration for a name occurs
+     * @param commitment The commitment calculated by makeCommitment
+     */
+    function commit(bytes32 commitment) public {
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:164]: not needed

File: contracts/contracts/Tweet.sol [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+        dc.commit(commitment);
+    }
+
+    function getPrice(string memory name) public view returns (uint256) {
+        uint256 ensPrice = dc.getENSPrice(name, duration);
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:169]: not needed

File: contracts/contracts/Tweet.sol [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+        uint256 price = getPrice(name);
+
+        require(price <= msg.value, "DC: insufficient payment");
+        require(available(name), "DC: name unavailable");
+        dc.register{value: ensPrice}(name, owner, duration, secret);
+        uint256 tokenId = uint256(keccak256(bytes(name)));
+        NameRecord storage nameRecord = nameRecords[bytes32(tokenId)];
+        nameRecord.renter = owner;
+        nameRecord.lastPrice = price;
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:200]: no renter

File: contracts/contracts/Tweet.sol [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+        emit NameRenewed(name, nameRecord.renter, price, url);
+
+        uint256 excess = msg.value - price;
+        if (excess > 0) {
+            (bool success, ) = msg.sender.call{value: excess}("");
+            require(success, "cannot refund excess");
+        }
+    }
+
+    function getReinstateCost(string calldata name) public view returns (uint256){
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:252]: not needed

File: contracts/contracts/Tweet.sol [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+        if (excess > 0) {
+            (bool success, ) = msg.sender.call{value: excess}("");
+            require(success, "cannot refund excess");
+        }
+    }
+
+    modifier recordOwnerOnly(string calldata name){
+        NameRecord storage r = nameRecords[keccak256(bytes(name))];
+        require(dc.ownerOf(name) == msg.sender, "DC: not nameWrapperowner");
+        require(r.renter == msg.sender, "DC: not owner");
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:297]: wrong

File: contracts/contracts/Tweet.sol [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+        return ret;
+    }
+
+    function withdraw() external {
+        require(msg.sender == owner() || msg.sender == revenueAccount, "DC: must be owner or revenue account");
+        (bool success, ) = revenueAccount.call{value: address(this).balance}("");
+        require(success, "DC: failed to withdraw");
+    }
+
+    receive() external payable{
```
Comment (polymorpher) [inline contracts/contracts/Tweet.sol:353]: should remove

File: contracts/contracts/DC.sol [hunk 16]
```diff
@@ -6,135 +6,72 @@ import "@openzeppelin/contracts/access/Ownable.sol";
...
     @notice This contract allows the rental of domains under .country (”DC”)
     it integrates with the ENSRegistrarController and the ENS system as a whole for persisting of domain registrations.
-    It is responsible for holding the revenue from these registrations for the web2 portion of the registration process,
-    with the web3 registration revenue being held by the RegistrarController contract.
-    An example would be as follows Alice registers alice.com and calls the register function with an amount of 10,000 ONE.
-    5000 ONE would be held by the DC contract and the remaining 5000 funds would be sent to the RegistrarController using 
-    the register function.
+    The calling services are responsible for holding the revenue from these registrations for the web2 portion of the 
+    registration process, with the web3 registration revenue being held by the RegistrarController contract.
+    An example would be as follows Alice registers alice.com this calls the service register function with an amount of 10,000 ONE.
```
Comment (polymorpher) [inline contracts/contracts/DC.sol:19]: vague, need clarification and definition

File: contracts/contracts/DC.sol [hunk 17]
```diff
@@ -6,135 +6,72 @@ import "@openzeppelin/contracts/access/Ownable.sol";
...
-    uint256 public duration;
     address public resolver;
     bool public reverseRecord;
     uint32 public fuses;
     uint64 public wrapperExpiry;
     bool public initialized;
 
     struct InitConfiguration {
-        uint256 baseRentalPrice;
-        uint256 duration;
```
Comment (polymorpher) [inline contracts/contracts/DC.sol:38]: should keep

File: contracts/contracts/DC.sol [hunk 18]
```diff
@@ -6,135 +6,72 @@ import "@openzeppelin/contracts/access/Ownable.sol";
...
 
-    mapping(bytes32 => NameRecord) public nameRecords;
-    mapping(bytes32 => string[]) public urlsPerRecord; // additional urls per record
-    string public lastRented;
-
-    bytes32[] public keys;
-
     event NameRented(string indexed name, address indexed renter, uint256 price, string url);
     event NameRenewed(string indexed name, address indexed renter, uint256 price, string url);
     event NameReinstated(string indexed name, address indexed renter, uint256 price, address oldRenter);
```
Comment (polymorpher) [inline contracts/contracts/DC.sol:51]: not needed

File: contracts/contracts/DC.sol [hunk 19]
```diff
@@ -161,27 +98,14 @@ contract DC is Pausable, Ownable {
...
-    }
-
     /**
      * @dev `available` calls RegistrarController to check if a name is available
      * @param name The name to be checked being registered
      */
     function available(string memory name) public view returns (bool) {
-        NameRecord storage record = nameRecords[keccak256(bytes(name))];
-        return registrarController.available(name) && (record.renter == address(0) || uint256(record.expirationTime) + gracePeriod <= block.timestamp);
+        // NameRecord storage record = nameRecords[keccak256(bytes(name))];
```
Comment (polymorpher) [inline contracts/contracts/DC.sol:106]: remove comments

File: contracts/contracts/DC.sol [hunk 20]
```diff
@@ -191,7 +115,7 @@ contract DC is Pausable, Ownable {
      * @param owner The address of the owner of the name being registered
      * @param secret A random secret passed by the client
      */
-    function makeCommitment(string memory name, address owner, bytes32 secret) public view returns (bytes32) {
+    function makeCommitment(string memory name, address owner, uint256 duration, bytes32 secret) public view returns (bytes32) {
```
Comment (polymorpher) [inline contracts/contracts/DC.sol:118]: should pre-set duration

File: contracts/contracts/DC.sol [hunk 21]
```diff
@@ -209,49 +133,25 @@ contract DC is Pausable, Ownable {
      * tRegistrarController.rentPrice (price.base + price.premium)
      * @param name The name being registered
      */
-    function getENSPrice(string memory name) public view returns (uint256) {
+    function getENSPrice(string memory name, uint256 duration) public view returns (uint256) {
```
Comment (polymorpher) [inline contracts/contracts/DC.sol:136]: pre-set duration

File: contracts/contracts/DC.sol [hunk 22]
```diff
@@ -209,49 +133,25 @@ contract DC is Pausable, Ownable {
...
      * @dev `register` calls RegistrarController register and is used to register a name
      * this also takes a fee for the web2 registration which is held by DC.sol a check is made to ensure the value sent is sufficient for both fees
      * @param name The name to be registered e.g. for test.country it would be test
-     * @param url A URL that can be embedded in a web2 default domain page e.g. a twitter post
+     # @param owner The owner of the registerd name
+     * @param duration Length of time to register the name
      * @param secret A random secret passed by the client
      */
-    function register(string calldata name, string calldata url, bytes32 secret) public payable whenNotPaused {
+    function register(string calldata name, address owner, uint256 duration, bytes32 secret) public payable whenNotPaused {
```
Comment (polymorpher) [inline contracts/contracts/DC.sol:149]: pre-set duration

File: contracts/contracts/DC.sol [hunk 23]
```diff
@@ -266,147 +166,55 @@ contract DC is Pausable, Ownable {
...
-        }
-        if (expiration > nameRecord.expirationTime) {
-            chargeableDuration = expiration - nameRecord.expirationTime;
-        }
-        uint256 charge = (chargeableDuration * 1e18 / duration * baseRentalPrice) / 1e18;
-        return charge;
+    function verifyOnwer(string calldata name, address owner) public view returns(bool) {
+        bytes32 node = keccak256(bytes(name));
+        uint256 tokenId = uint256(node);
+        address currentOwner = nameWrapper.ownerOf(tokenId);
```
Comment (polymorpher) [inline contracts/contracts/DC.sol:197]: should check baseRegistrar too

File: contracts/contracts/DC.sol [hunk 24]
```diff
@@ -266,147 +166,55 @@ contract DC is Pausable, Ownable {
...
-    }
-
-    function getAllUrls(string calldata name) public returns (string[] memory){
-        bytes32 key = keccak256(bytes(name));
-        string[] memory ret = new string[](urlsPerRecord[key].length);
-        for (uint256 i = 0; i < urlsPerRecord[key].length; i++) {
-            ret[i] = urlsPerRecord[key][i];
-        }
-        return ret;
+    function ownerOf(string calldata name) public view returns(address) {
```
Comment (polymorpher) [inline contracts/contracts/DC.sol:207]: should check baseRegistrar too

File: contracts/contracts/DC.sol [hunk 25]
```diff
@@ -266,147 +166,55 @@ contract DC is Pausable, Ownable {
...
-        uint256 chargeableDuration = 0;
-        if (nameRecord.expirationTime == 0) {
-            chargeableDuration = expiration - block.timestamp;
-        }
-        if (expiration > nameRecord.expirationTime) {
-            chargeableDuration = expiration - nameRecord.expirationTime;
-        }
-        uint256 charge = (chargeableDuration * 1e18 / duration * baseRentalPrice) / 1e18;
-        return charge;
+    function verifyOnwer(string calldata name, address owner) public view returns(bool) {
```
Comment (polymorpher) [inline contracts/contracts/DC.sol:194]: seems redundant given ownerOf

File: contracts/contracts/DC.sol [hunk 26]
```diff
@@ -6,135 +6,72 @@ import "@openzeppelin/contracts/access/Ownable.sol";
...
-    address public revenueAccount;
     IRegistrarController public registrarController;
+    INameWrapper public nameWrapper;
     IBaseRegistrar public baseRegistrar;
-    uint256 public duration;
     address public resolver;
     bool public reverseRecord;
     uint32 public fuses;
     uint64 public wrapperExpiry;
     bool public initialized;
```
Comment (polymorpher) [inline contracts/contracts/DC.sol:32]: not needed since you removed initialize function

File: contracts/deploy/ensDC.ts [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+
+  const readConfiguration = {
+    wrapperExpiry: (await dc.wrapperExpiry()).toString(),
+    fuses: await dc.fuses(),
+
+    registrarController: await dc.registrarController(),
+    baseRegistrar: await dc.baseRegistrar(),
+    resolver: await dc.resolver(),
+    reverseRecord: await dc.reverseRecord(),
+    initialized: await dc.initialized()
```
Comment (polymorpher) [inline contracts/deploy/ensDC.ts:42]: remove

File: contracts/deploy/ensTweet.ts [hunk 28]
```diff
@@ -3,76 +3,62 @@ import { DeployFunction } from 'hardhat-deploy/types'
...
-  console.log('DC address:', DC.address)
-  const dc = await ethers.getContractAt('DC', DC.address) as DC
+  console.log('Tweet address:', Tweet.address)
+  const tweet = await ethers.getContractAt('Tweet', Tweet.address) as Tweet
   if (config.initialRecordFile) {
-    const records:{name:string, key: string, record: DC.NameRecordStruct}[] = JSON.parse(await fs.readFile(config.initialRecordFile, { encoding: 'utf-8' }))
+    const records:{name:string, key: string, record: Tweet.NameRecordStruct}[] = JSON.parse(await fs.readFile(config.initialRecordFile, { encoding: 'utf-8' }))
     const chunks = chunk(records, 50)
-    await dc.initialize([ethers.utils.id('')], [{ renter: ethers.constants.AddressZero, lastPrice: '0', rentTime: '0', expirationTime: '0', url: '', prev: '', next: chunks[0][0].name }])
+    await tweet.initialize([ethers.utils.id('')], [{ renter: ethers.constants.AddressZero, lastPrice: '0', rentTime: '0', expirationTime: '0', url: '', prev: '', next: chunks[0][0].name }])
```
Comment (polymorpher) [inline contracts/deploy/ensTweet.ts:35]: no renter

File: contracts/deploy/ensTweet.ts [hunk 29]
```diff
@@ -3,76 +3,62 @@ import { DeployFunction } from 'hardhat-deploy/types'
...
     from: deployer,
     args: [initConfiguration],
     log: true,
     autoMine: true // speed up deployment on local network (ganache, hardhat), no effect on live networks
   })
-  console.log('DC address:', DC.address)
-  const dc = await ethers.getContractAt('DC', DC.address) as DC
+  console.log('Tweet address:', Tweet.address)
+  const tweet = await ethers.getContractAt('Tweet', Tweet.address) as Tweet
   if (config.initialRecordFile) {
```
Comment (polymorpher) [inline contracts/deploy/ensTweet.ts:32]: incompatible

Comment (polymorpher) [reply contracts/deploy/ensTweet.ts:32]: should remove init block

File: contracts/deploy/ensTweet.ts [hunk 30]
```diff
@@ -3,76 +3,62 @@ import { DeployFunction } from 'hardhat-deploy/types'
...
       })
       console.log(`initializing ${names.length} records, starting from ${names[0]}: ${JSON.stringify(records[0])}`)
-      await dc.initialize(names, records)
+      await tweet.initialize(names, records)
     }
   }
-  await dc.finishInitialization()
-  const n = (await dc.numRecords()).toNumber()
-  console.log(`D1DC finished initialization for ${n} records`)
+  await tweet.finishInitialization()
```
Comment (polymorpher) [inline contracts/deploy/ensTweet.ts:46]: remove

**Comment by polymorpher** [comment]
merging to side branch, will edit myself

**Comment by polymorpher** [comment]
See https://github.com/polymorpher/dot-country/pull/4/files

**Pull request closed without merging**